internal/store/store.go

```diff
@@ -463,6 +516,75 @@ func (s *ObjectStore) referencedList(ctx context.Context, list model.ObjectList,
 	return nil
 }
 
+func (s *ObjectStore) updateForeignKeysTx(
+	ctx context.Context,
+	obj model.Object,
+	indexes model.Indexes,
+	tx persistence.Tx,
+) error {
+	// Ensure the "parent" entity exists. This is also used to resolve the full object for the caller's use.
+	if err := s.readByTypeID(ctx, tx, obj.Type(), obj.ID(), obj); err != nil {
+		return err
+	}
+
+	// Create/remove foreign keys as requested. This is handled asynchronously
+	// as bulk management is supported & this should be a performant call.
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := range indexes {
+		index := indexes[i]
+		g.Go(func() error {
+			key, value, err := s.indexKV(index, obj)
+			if err != nil {
+				return fmt.Errorf("unable to render indexes: %w", err)
+			}
+
+			if index.Action == model.IndexActionAdd {
+				// Ensure the foreign entity exists.
+				fk, err := s.genID(index.ForeignType, index.Value)
+				if err != nil {
+					return err
+				}
+				if _, err = tx.Get(gCtx, fk); err != nil {
+					if errors.As(err, &persistence.ErrNotFound{}) {
+						return ErrConstraint{Index: index}
+					}
+					return err
+				}
+
+				if err := tx.Insert(gCtx, key, value); err != nil {
+					return err
+				}
```

Comment:
In this case, you can map the ErrUniqueViolation error to a ErrConstraint because you have that information, like the above if statement.

Reply:
Yeah, you're right. Went ahead and updated this.

Reply:
This has been fixed up.